### ISSUE TITLE
"Ease" pen up and down 

### ISSUE THREAD

**Issue by artBoffin** [evil-mad/axidraw#56]
Suggestion for an advanced feature to "ease" a pen up and down, that is - when movement starts the pen is up and moves to the down position over X amount of time (or distance). Same with the end of stroke, where the pen starts to move back up before the stroke finishes.

This is specifically for brush pens and flexible tip markers where having a thin to thick stroke is desirable at the start and end of stroke. Coupled with the rubber band, it could make for interesting styles.

Another possibility is a random (or defined by an f(x,y,t) function) "wiggle" movement while drawing to make variable line width.

**Comment by oskay** [comment]
This is essentially a request for 3D movements, which are not something that we're currently planning to support in cases where we're plotting from an SVG file.

The AxiDraw is primarily intended as a "2.1 axis" or "2.5 axis" machine, where you set a given Z height, and then work at that level with XY motion. However, there are a couple of ways to get around this, if you are directly commanding the machine, rather than plotting a 2D file. 

First, you can send a pen lift (or lower) command directly before an XY move command. For example, the command `SP,1\rSM,1000,0,3200` will begin the Z move and then (almost immediately) begin the XY move, such that the two happen simultaneously in practice. Note that this use of the SP command does not specify a delay time. If you do specify a delay time (as in `SP,1,1000\rSM,1000,0,3200`), then the XY move will not begin until the Z move completes. The rate of the pen lift can be specified with the SC command.

It would be relatively straightforward to add this kind of thing to the beginning of each pen-down movement, so that the XY movement begins before the pen hits the paper, but it would not be consistent, since the AxiDraw is not designed to provide a consistent pen-over-paper height. 

A second approach (perhaps more suited to the types of generative work that you are doing), you can directly and frequently change the definition of (for example) the pen-down Z position, by giving a command like `SC,4,10000\rSP,1` that both configures the pen down position and moves to that new position. Note that the value of Z position that you send (10000 here) should be kept in the range specified in axidraw_conf.py, 7500 to 28000, in order to prevent damage to the mechanism. This type of command also be executed before an XY move, to give the same type of apparent simultaneous motion.

**Comment by ammonialime** [comment]
@oskay Is there any way to get this ease effect with the python library?

**Comment by oskay** [comment]
You can use those very commands that I mentioned within the AxiDraw Python API, if thats what you're asking.

**Comment by elephantsneverforget** [comment]
The docs say using the direct USB commands isn't recommended. Is this the only way to get a moving Z axis while X/Y movements are underway?

**Comment by oskay** [comment]
The "easy" way is to just change [`pen_delay_down`](https://axidraw.com/doc/py_api/#pen_delay_down) and [`pen_delay_up`](https://axidraw.com/doc/py_api/#pen_delay_up).

**Comment by elephantsneverforget** [comment]
Thank you. Very helpful.

I see commands can be tied to [layers](https://wiki.evilmadscientist.com/AxiDraw_Layer_Control). But I don't see `pen_delay_down` or `pen_delay_up` in this list of commands. I assume this means it's not possible to get similar behaviour in Inkscape?

**Comment by oskay** [comment]
What does the layer control and/or Inkscape have to do with this question? What are you actually trying to do here?

Normally one uses *either* the Python API or Inkscape; it isn't really possible to use both at the same time.

**Comment by elephantsneverforget** [comment]
I'm trying to get variation in a brush stroke by moving the brush up/down during the drawing process. I understand how to do this in the Python interactive environment from your answer above. I'm wondering if it's also possible to get this behaviour in Inkscape.

**Comment by oskay** [comment]
Yes, it's available in Inkscape as well; it's not related to the layer controls.
